tools/nl/embeddings/build_embeddings_v2.py

```diff
@@ -0,0 +1,286 @@
+
+
+def _merge_dataframes(df_1: pd.DataFrame, df_2: pd.DataFrame) -> pd.DataFrame:
+  return df_1.merge(df_2, how='left', on=COL_ID).fillna("")
+
+
+def _concat_alternatives(alternatives: List[str],
+                         max_alternatives,
+                         delimiter=";") -> str:
+  alts = set(alternatives[0:max_alternatives])
+  return f"{delimiter}".join(alts)
+
+
+def _split_alt_string(alt_string: str) -> List[str]:
+  alts = []
+  for alt in alt_string.strip().split(";"):
+    if alt and alt != ";":
+      alts.append(alt)
+  return alts
+
+
+def _build_embeddings(ctx, texts: List[str], dcids: List[str]) -> pd.DataFrame:
+  embeddings = ctx.model.encode(texts, show_progress_bar=True)
+  embeddings = pd.DataFrame(embeddings)
+  embeddings[COL_ID] = dcids
+  embeddings[COL_ALTERNATIVES] = texts
+  return embeddings
+
+
+def get_sheets_data(ctx, sheets_url: str, worksheet_name: str) -> pd.DataFrame:
+  sheet = ctx.gs.open_by_url(sheets_url).worksheet(worksheet_name)
+  df = pd.DataFrame(sheet.get_all_records()).fillna("")
+  if SHEETS_DCID_COL in df.columns:
+    df.rename(columns={SHEETS_DCID_COL: COL_ID}, inplace=True)
+  return df
+
+
+def get_local_alternatives(local_filename: str,
+                           local_col_names: List[str]) -> pd.DataFrame:
+  df = pd.read_csv(local_filename).fillna("")
+  df = df[local_col_names]
+  if CSV_DCID_COL in df.columns:
+    df.rename(columns={CSV_DCID_COL: COL_ID}, inplace=True)
+  return df
+
+
+def get_embeddings(ctx, df_svs: pd.DataFrame,
+                   local_merged_filepath: str) -> pd.DataFrame:
+  print(f"Concatenate all alternative sentences for descriptions.")
+  alternate_descriptions = []
+  for _, row in df_svs.iterrows():
+    alternatives = []
+    if row[SHEETS_OVERRIDE_COL]:
+      # Override takes precendence over everything else.
+      alternatives += _split_alt_string(row[SHEETS_OVERRIDE_COL])
+    else:
+      for col_name in [
+          SHEETS_NAME_COL,
+          SHEETS_DESCRIPTION_COL,
+          SHEETS_ALTERNATIVES_COL,
+          CSV_OTHER_ALTERNATIVES_COL,
+          CSV_PALM_ALTERNATIVES_COL,
+      ]:
+        # In order of preference, traverse the various alternative descriptions.
+        alternatives += _split_alt_string(row[col_name])
+
+    alt_str = _concat_alternatives(alternatives, MAX_ALTERNATIVES_LIMIT)
+    alternate_descriptions.append(alt_str)
+
+  assert len(df_svs) == len(alternate_descriptions)
+  df_svs[COL_ALTERNATIVES] = alternate_descriptions
+
+  # Write to local_merged_filepath.
+  print(
+      f"Writing the concatenated dataframe after merging alternates to local file: {local_merged_filepath}"
+  )
+  df_svs[[COL_ID, COL_ALTERNATIVES]].to_csv(local_merged_filepath, index=False)
+
+  # Build embeddings.
+  print("Getting texts, dcids and embeddings.")
+  df_svs = _trim_columns(df_svs)
+  (texts, dcids) = _get_texts_dcids(df_svs)
+
```

Comment:
nit: maybe _trim_columns first so we can just write the df_svs to local path?

Reply:
my thought was to keep everything in the local file and the trim() only really helps for the embeddings.